File: src/main/java/dev/jlo/gamemodes/paper/CommandRouter.java
package dev.jlo.gamemodes.paper;

import java.util.List;
import java.util.UUID;

/** Pure command parser shared by Bukkit executor and fake-based tests. */
public final class CommandRouter {
    private final CommandService service;

    public CommandRouter(CommandService service) {
        this.service = service;
    }

    public String execute(UUID player, List<String> arguments) {
        if (arguments.isEmpty()) {
            return "Usage: join <opr|siege> [arena], leave, ready, status, team";
        }
        String command = arguments.getFirst().toLowerCase();
        return switch (command) {
            case "opr", "siege" -> join(player, Mode.parse(command), getOrNull(arguments, 1));
            case "join" -> {
                String modeValue = getOrNull(arguments, 1);
                Mode mode = modeValue == null ? null : Mode.parse(modeValue);
                if (mode == null) {
                    yield "Usage: join <opr|siege> [arena]";
                }
                yield join(player, mode, getOrNull(arguments, 2));
            }
            case "leave" -> service.leave(player) ? "left" : "not in a queue or match";
            case "ready" -> service.ready(player) ? "ready" : "not in a queue or match";
            case "status" -> service.status(player);
            case "team" -> {
                TeamAssignment assignment = service.team(player);
                yield assignment != null ? "team " + assignment.name() : "not in a queue or match";
            }
            default -> "Unknown command: " + command;
        };
    }

    private String join(UUID player, Mode mode, String arena) {
        service.join(player, mode, arena);
        return "joined " + mode.name();
    }

    private static <T> T getOrNull(List<T> values, int index) {
        return index < values.size() ? values.get(index) : null;
    }
}

File: src/main/kotlin/dev/jlo/gamemodes/paper/CommandRouter.kt
package dev.jlo.gamemodes.paper

import java.util.UUID

interface CommandService {
    fun join(player: UUID, mode: Mode, arena: String? = null): TeamAssignment
    fun leave(player: UUID): Boolean
    fun ready(player: UUID): Boolean
    fun status(player: UUID): String
    fun team(player: UUID): TeamAssignment?
    fun admin(player: UUID?, action: String, args: List<String>): String
}

/** Pure command parser shared by Bukkit executor and fake-based tests. */
class CommandRouter(private val service: CommandService) {
    fun execute(player: UUID, arguments: List<String>): String {
        if (arguments.isEmpty()) return "Usage: join <opr|siege> [arena], leave, ready, status, team"
        val command = arguments.first().lowercase()
        return when (command) {
            "opr", "siege" -> join(player, Mode.parse(command)!!, arguments.getOrNull(1))
            "join" -> {
                val mode = arguments.getOrNull(1)?.let(Mode::parse)
                    ?: return "Usage: join <opr|siege> [arena]"
                join(player, mode, arguments.getOrNull(2))
            }
            "leave" -> if (service.leave(player)) "left" else "not in a queue or match"
            "ready" -> if (service.ready(player)) "ready" else "not in a queue or match"
            "status" -> service.status(player)
            "team" -> service.team(player)?.let { "team ${it.name}" } ?: "not in a queue or match"
            else -> "Unknown command: $command"
        }
    }

    private fun join(player: UUID, mode: Mode, arena: String?): String =
        "joined $mode".replace("$mode", mode.name)
            .let { service.join(player, mode, arena); it }
}
